src/chk/chk_leader.c

```diff
@@ -277,20 +279,1032 @@ chk_leader_find_slowest(struct chk_instance *ins)
+	 */
+	if (*result == 0 || !(ins->ci_prop.cp_flags & CHK__CHECK_FLAG__CF_FAILOUT)) {
+		if (notify) {
+			iv.ci_gen = cbk->cb_gen;
+			uuid_copy(iv.ci_uuid, uuid);
+			iv.ci_phase = cbk->cb_phase;
+			if (*result != 0)
+				iv.ci_status = CHK__CHECK_INST_STATUS__CIS_FAILED;
+			else
+				iv.ci_status = CHK__CHECK_INST_STATUS__CIS_IMPLICATED;
+			iv.ci_remove_pool = 1;
+
+			/* Synchronously notify the engines that check on the pool got failure. */
+			rc = chk_iv_update(ins->ci_iv_ns, &iv, CRT_IV_SHORTCUT_NONE,
+					   CRT_IV_SYNC_EAGER, true);
+			if (rc != 0)
+				D_ERROR(DF_LEADER" failed to notify the engines that "
+					"the pool "DF_UUID" got failure: "DF_RC"\n",
+					DP_LEADER(ins), DP_UUID(uuid), DP_RC(rc));
+		}
+
+		if (update)
+			chk_bk_update_leader(cbk);
+
+		*result = 0;
+	}
+}
+
+static d_rank_list_t *
+chk_leader_cpr2ranklist(struct chk_pool_rec *cpr, bool svc)
+{
+	struct chk_pool_shard	*cps;
+	struct ds_pool_clue	*clue;
+	d_rank_list_t		*ranks;
+	int			 i = 0;
+
+	ranks = d_rank_list_alloc(cpr->cpr_shard_nr);
+	if (ranks != NULL) {
+		d_list_for_each_entry(cps, &cpr->cpr_shard_list, cps_link) {
+			if (svc) {
+				clue = cps->cps_data;
+				if (clue == NULL || clue->pc_svc_clue == NULL)
+					continue;
+			}
+			ranks->rl_ranks[i++] = cps->cps_rank;
+		}
+
+		ranks->rl_nr = i;
+	}
+
+	return ranks;
+}
+
+static bool
+chk_pool_in_zombie(struct chk_pool_rec *cpr)
+{
+	struct chk_pool_shard	*cps;
+	struct ds_pool_clue	*clue;
+	bool			 found = false;
+
+	d_list_for_each_entry(cps, &cpr->cpr_shard_list, cps_link) {
+		clue = cps->cps_data;
+		if (clue->pc_dir == DS_POOL_DIR_ZOMBIE) {
+			found = true;
+			break;
+		}
+	}
+
+	return found;
+}
+
 static int
-chk_leader_handle_pools_p1(struct chk_sched_args *csa)
+chk_leader_destroy_pool(struct chk_pool_rec *cpr, uint64_t seq, bool dereg)
 {
-	/* TBD: merge with Liwei's patch. */
+	d_rank_list_t	*ranks = NULL;
+	int		 rc = 0;
 
-	return 0;
+	/*
+	 * XXX: Firstly, deregister from MS. If it is successful but we failed to destroy
+	 *	related pool target(s) in subsequent steps, then the pool becomes orphan.
+	 *	It may cause some space leak, but will not cause correctness issue. That
+	 *	will be handled when run DAOS check next time.
+	 */
+	if (dereg) {
+		rc = ds_chk_deregpool_upcall(seq, cpr->cpr_uuid);
+		if (rc != 0)
+			goto out;
+	}
+
+	ranks = chk_leader_cpr2ranklist(cpr, false);
+	if (ranks == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rc = ds_mgmt_tgt_pool_destroy(cpr->cpr_uuid, ranks);
+	d_rank_list_free(ranks);
+
+out:
+	return rc;
+}
+
+static struct ds_pool_clue *
+chk_leader_locate_pool_clue(struct chk_pool_rec *cpr)
+{
+	struct chk_pool_shard	*cps;
+	struct ds_pool_clue	*clue = NULL;
+	int			 i = 0;
+
+	D_ASSERT(cpr->cpr_advice >= 0);
+	D_ASSERT(cpr->cpr_advice < cpr->cpr_shard_nr);
+
+	d_list_for_each_entry(cps, &cpr->cpr_shard_list, cps_link) {
+		clue = cps->cps_data;
+		if (clue == NULL || clue->pc_svc_clue == NULL)
+			continue;
+
+		if (i++ == cpr->cpr_advice)
```

Comment:
The advice returned by ds_pool_check_svc_clues is an index into the clues parameter, which only includes clues for shards with PS replicas. Since cpr_shard_list may include shards without PS replicas (am I mistaken?), we can't use the advice on cpr_shard_list.

Reply:
If the previous comment is not mistaken, we could consider changing cpr_advice to be a d_rank_t.

Reply:
Does cpr_shard_list always comprise the same list of ranks as cpr_clues?

Reply:
Ah, okay, I see. The key is the if statement. (I'd probably not code this way, for the dependency across the two functions is too subtle, especially because there are better, simpler ways.)